### PR TITLE
dbeaver/dbeaver#23070 Print the driver that isn't downloaded

### LINKED ISSUE #23070: Show the URL attempted when connection not working
**Is your feature request related to a problem? Please describe.**
I'm using DBeaver behind a restrictive proxy, and when it tries to download drivers, it gets blocked and just simply shows the error that it couldn't connect.

**Describe the solution you'd like**
If it can at least show me what URL it was trying then I can submit that to the proxy bypass to get it to work.

**Describe alternatives you've considered**
The alternative to download the driver actually did not work. I was able to get the driver installed, but yet, it still is stuck on some internet connection, and again, I have no idea what it's trying to connect to.

### PR DESCRIPTION
**Pull request by HocKu7**
Now we show information about the failed downloading and, particularly, the URL to jar that we didn't manage to download
image

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class ThreadExceptionsContext {
+
+    private static final Map<Thread, List<Exception>> THREAD_TO_EXCEPTIONS = new ConcurrentHashMap<>();
+    private static final Map<Thread, Thread> THREAD_BINDING_MAP = new ConcurrentHashMap<>();
+    //private static final Logger log = LoggerFactory.getLogger(ThreadExceptionsContext.class);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]:
```suggestion
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    public static List<Exception> getListExceptionsForCurrentThread() {
+        Thread threadForRegistry = getThreadForRegistry();
+        List<Exception> exceptionList = THREAD_TO_EXCEPTIONS.get(threadForRegistry);
+        if(exceptionList == null){
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]:
```suggestion
        if (exceptionList == null) {
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return new ArrayList<>(exceptionList);
+    }
+
+    public static Set<Exception> getSetExceptionsForCurrentThread() {
+        return new HashSet<>(getListExceptionsForCurrentThread());
+    }
+
+    public static void registerExceptionForCurrentThread(Exception exception) {
+
+        Thread threadForRegistry = getThreadForRegistry();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]:
```suggestion
    public static void registerExceptionForCurrentThread(Exception exception) {
        Thread threadForRegistry = getThreadForRegistry();
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            //log.warn(String.format("Thread %s is already in the binding map for ThreadExceptionContext. Overriding...",
+            //    currentThread.getName()));
+        }
+        THREAD_BINDING_MAP.put(currentThread, callerThread);
+    }
+
+
+    public static void realiseResources() {
+
+        Thread currentThread = Thread.currentThread();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]:
```suggestion
    public static void realiseResources() {
        Thread currentThread = Thread.currentThread();
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java [hunk 5]
```diff
@@ -579,6 +579,36 @@ public static IStatus makeExceptionStatus(int severity, Throwable ex) {
         return makeExceptionStatus(severity, ex, false);
     }
 
+    public static IStatus transformExceptionsToStatus(@NotNull Set<Exception> exceptions) {
+
+        if (exceptions.isEmpty()) {
+            return new Status(IStatus.ERROR, (Class<?>) null, "Empty exceptions list");
+        }
+        Set<String> excaptionMessageSet = new HashSet<>();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java]:
```suggestion
        Set<String> exceptionMessageSet = new HashSet<>();
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     * If the thread which you are registering is a different thread that gonna read exception list
+     *
+     * @param callerThread a key where you will be able to find exceptions for current thread
+     */
+    public static void bindCurrentContextFor(Thread callerThread) {
+        Thread currentThread = Thread.currentThread();
+        //todo if it will need, there can be used a set for threads
+        if (THREAD_BINDING_MAP.get(currentThread) != null) {
+            log.warn(String.format("Thread %s is already in the binding map for ThreadExceptionContext. Overriding...",
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]: Do we really need to log it? User might see his message, but it's hard to understand, what this message is about, without context of this code

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverLibraryMavenArtifact.java [hunk 7]
```diff
@@ -339,7 +339,8 @@ protected MavenArtifactVersion resolveLocalVersion(DBRProgressMonitor monitor, b
         }
         MavenArtifactVersion version = getArtifactVersion(monitor);
         if (version == null) {
-            throw new IOException("Maven artifact '" + path + "' not found");
+            String versionMessageError = preferredVersion != null ? String.format(":%s", preferredVersion) : "";
+            throw new IOException(String.format("Maven artifact %s %s not found", path, versionMessageError));
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverLibraryMavenArtifact.java]:
```suggestion
            throw new IOException(String.format("Maven artifact %s %s is not found", path, versionMessageError));
```

**Comment by ShadelessFox** [comment]
I don't understand the whole logic behind this PR. What does it do? Stores a list of exceptions per thread?

**Comment by E1izabeth** [comment]
Yes, as far as I understand, it accumulated exceptions per thread between the first download attempt and the end of attempting. And I agree, that it's doubtful

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java (hunk 6 above)
Comment (HocKu7) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]: I suppose there could be a concurrency issue that we can't see right now, so I prefer to see any such bugs related to concurrency in the log. I'm thinking of the log for developers, not for users, isn't it?

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ThreadExceptionsContext.java]: Yes, but sometimes users read logs, if some bug occurred. 
If it's an exclusive situation, then I'm fine with logging it. I thought it will be called each time.

**Comment by HocKu7** [comment]
Generally speaking, the main goal of the PR is to show the URL of the driver that we didn't manage to reach. For this, I created a ThreadExceptionContextx that allows us to collect exceptions through different threads. Otherwise, there is a need a complicated refactoring

**Comment by ShadelessFox** [comment]
Maybe we need refactoring. This solution is quite complex and really hard to follow/debug.

**Comment by ShadelessFox** [comment]
Also, on your screenshot, there are two mentions of the same file but in different repositories, but the top message clearly says that DBeaver can't access the last one. Is this expected?

**Comment by HocKu7** [comment]
We discussed that with @serge-rider. He would like to have some variation of the Exception Context.  About the screenshot - yes, it corresponds to the current logic of the attempt to download the driver. It tries to download the driver from each repository(maven, sonatype), and then if it gets an exception, it checks internet connectivity to maeven.org and gets an exception again. So, there are three exceptions

**Comment by serge-rider** [comment]
I don't like the approach with thread local variables.
It is better to pass ErrorContext to WebUtils and other methods as a parameter

**Comment by ShadelessFox** [comment]
Why don't you use the existing mechanism of suppressed exceptions? When catching multiple exceptions, collect them and add them as suppressed to a parent exception

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/WebUtils.java [hunk 8]
```diff
@@ -56,6 +58,11 @@ public static URLConnection openConnection(String urlString, DBPAuthInfo authInf
         return openURLConnection(urlString, authInfo, referrer, 1);
     }
 
+    @NotNull
+    public static URLConnection openConnection(String urlString, DBPAuthInfo authInfo, String referrer, DBRProgressMonitor monitor) throws IOException {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/WebUtils.java]: By convention monitor is always the first parameter

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverDependenciesTree.java [hunk 9]
```diff
@@ -192,16 +200,24 @@ private void grayOutInstalledArtifact(DBPDriverDependencies.DependencyNode node,
...
 
-    public boolean handleDownloadError(DBException e) {
+    public boolean handleDownloadError(DBException causeException) {
         try {
             checkNetworkAccessible();
         } catch (DBException dbException) {
-            DBWorkbench.getPlatformUI().showError("Download error",
-                "Network error", dbException);
-            return false;
+            if(causeException instanceof DBExceptionWithHistory exceptionWithHistory){
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverDependenciesTree.java]: Code style

**Review by serge-rider — CHANGES_REQUESTED** [review]

**Comment by HocKu7** [comment]
> Why don't you use the existing mechanism of suppressed exceptions? When catching multiple exceptions, collect them and add them as suppressed to a parent exception

Decided that approach with monitor more suitable for the task

**Comment by Matvey16** [comment]
`Exception during to connect to` while understabdable from technical standpoint, is not a correct sentance. Can we change it?

**Comment by Matvey16** [comment]
Verified
